test/integration/auth/rbac_test.go

```diff
@@ -816,3 +819,320 @@ func TestDiscoveryUpgradeBootstrapping(t *testing.T) {
 		t.Errorf("`system:public-info-viewer` should have inherited Subjects from `system:discovery` Wanted: %v, got %v", newDiscRoleBinding.Subjects, publicInfoViewerRoleBinding.Subjects)
 	}
 }
+
+type authorizeRequest struct {
+	ctx      context.Context
+	ar       authorizer.AttributesRecord
+	expected authorizer.Decision
+}
+
+// For 1.31 ctx was wired into the authorizers. This tests check that context values
+// are not used inside the code to resolve namespaces or users with the goal of
+// preventing regressions in the future.
+func TestRBACContextContamination(t *testing.T) {
+	superUser := "admin/system:masters"
+
+	tests := []struct {
```

Comment:
Let's simplify this so the test server only has to start once

There's three combinations of API objects we need to test:

* clusterrole+clusterbinding
  * should allow cluster-scoped request
  * should allow namespace-scoped request in any namespace
  * should disallow request for resource not in rules
* clusterrole+rolebinding
  * should disallow cluster-scoped request
  * should allow namespace-scoped request in rolebinding namespace
  * should disallow namespace-scoped request in other namespace
  * should disallow request for resource not in rules
* role+rolebinding
  * should disallow cluster-scoped request
  * should allow namespace-scoped request in rolebinding namespace
  * should disallow namespace-scoped request in other namespace
  * should disallow request for resource not in rules

To test that, create each pair of API objects with a distinct name and user, then add the testcases for that user, something like this:

```go
roles := bootstrapRoles{}

// clusterrole + clusterrolebinding scenarios
roles.clusterRoles = append(roles.clusterRoles, rbacapi.ClusterRole{
  ObjectMeta: metav1.ObjectMeta{Name: "clusterrole-clusterrolebinding"},
  Rules:      []rbacapi.PolicyRule{ruleReadPods},
})
roles.clusterRoleBindings = append(roles.clusterRoleBindings, rbacapi.ClusterRoleBinding{
  ObjectMeta: metav1.ObjectMeta{Name: "clusterrole-clusterrolebinding"},
  Subjects:   []rbacapi.Subject{{Kind: "User", Name: "clusterrole-clusterrolebinding-user"}},
  RoleRef:    rbacapi.RoleRef{Kind: "ClusterRole", Name: "clusterrole-clusterrolebinding"},
})
clusterRoleClusterRoleBindingUser := &user.DefaultInfo{Name: "clusterrole-clusterrolebinding-user"}
testcases = append(testcases,
  {
    ar: authorizer.AttributesRecord{Verb: "list", Resource: "pods", Namespace: "", ResourceRequest: true, User: clusterRoleClusterRoleBindingUser},
    expected: authorizer.DecisionAllow,
  },
  {
    ar: authorizer.AttributesRecord{Verb: "list", Resource: "pods", Namespace: "valid", ResourceRequest: true, User: clusterRoleClusterRoleBindingUser},
    expected: authorizer.DecisionAllow,
  },
  {
    ar: authorizer.AttributesRecord{Verb: "list", Resource: "configmaps", Namespace: "", ResourceRequest: true, User: clusterRoleClusterRoleBindingUser},
    expected: authorizer.DecisionNoOpinion,
  },
)
```

Then instantiate the test server once and run through the testcases. For each testcase, pass in a context without a namespace, one with an empty namespace, and one with an invalid namespace, and ensure the passed-in context has no bearing on the expected result.

```go
for _, tc := range testcases {
  decision, _, err := rbacAuthz.Authorize(context.Background(), &r.ar)
  ...
  decision, _, err := rbacAuthz.Authorize(genericapirequest.WithNamespace(context.Background(), ""), &r.ar)
  ...
  decision, _, err := rbacAuthz.Authorize(genericapirequest.WithNamespace(context.Background(), "invalid"), &r.ar)
  ...
}
```

Reply:
Wired in the tests as per your suggestion. Just left biut more verbose (c1, c2, c3) as found it easier in logs when it fails (`User c2-user does not have permissions.....`) to identify what failed.

Added everywhere `genericapirequest.WithNamespace(ctx, "")` where new ctx was wired in and where it used in auth stack.

Reply:
@liggitt do you think this this can get in into this release?

Reply:
> Added everywhere genericapirequest.WithNamespace(ctx, "") where new ctx was wired in and where it used in auth stack.

Um, I'm not sure we should add this *everywhere*. I'm pretty sure doing so creates multiple nested calls where we're now stacking multiple nested context objects because of this. The context implementation creates a linked list of context objects, each holding a single value, where every access traverses the list looking for a value. Nesting like this makes every context lookup have to go through multiple levels to find the value it wants.

I think we should *only* do this bit in the AuthorizerAdapter types where we transition non-API calls into "fake" API calls into the registry:
* AuthorizerAdapter#GetClusterRole
* AuthorizerAdapter#ListClusterRoleBindings
* AuthorizerAdapter#GetRole
* AuthorizerAdapter#ListRoleBindings

Reply:
Missed this comment :facepalm: 
Updated!